pallets/liquidity-pools-gateway/src/lib.rs

```diff
@@ -530,9 +625,74 @@ pub mod pallet {
 			domain_address: DomainAddress,
 			message: T::Message,
 		) -> (DispatchResult, Weight) {
-			let weight = Weight::from_parts(0, T::Message::max_encoded_len() as u64)
+			let mut weight = Weight::from_parts(0, T::Message::max_encoded_len() as u64)
 				.saturating_add(LP_DEFENSIVE_WEIGHT);
 
+			let routers_count = match DomainMultiRouters::<T>::get(domain_address.domain()) {
+				Some(r) => r.len(),
+				None => return (Err(Error::<T>::MultiRouterNotFound.into()), weight),
+			};
+
+			weight = weight.saturating_add(T::DbWeight::get().reads(1));
+
+			if routers_count <= 1 {
+				return (Err(Error::<T>::NotEnoughRouters.into()), weight);
+			}
+
+			let expected_proof_count: u32 = (routers_count - 1) as u32;
+
+			let (message_proof, message_proof_count) = match message.get_message_proof() {
+				None => {
+					let message_proof = message
+						.to_message_proof()
+						.get_message_proof()
+						.expect("message proof ensured by 'to_message_proof'");
+
+					InboundMessages::<T>::insert(message_proof, message.clone());
+
+					(
+						message_proof,
+						InboundMessageProofCount::<T>::get(message_proof),
+					)
+				}
+				Some(message_proof) => {
+					let message_proof_count =
+						match InboundMessageProofCount::<T>::try_mutate(message_proof, |count| {
+							*count += 1;
```

Comment:
Could the same router send the proof 3 times and we will count 3?

Reply:
In this current state, yes. Especially given that we can get the same message more than once.

Reply:
Added some tests with different nonproof/proof messages to ensure that these storages remain sane, please let me know what you think.

Reply:
I'll check in the next review pass!